Updated library to fully use Task infrastructure for scheduled functions

tomlm opened the pull request:
The usage pattern had scheduled tasks as sync functions, which I believe to be a leftover from the pattern of the original library from 2010. 

By changing the IScheduledTask function to return Task<bool> then scheduled functions can simply be sync or async cleanly. This also allows for much cleaner semantics around cancellation token. 
For example:
```C#
   schedule
       .AtMinutes(0,15,30,45)
       .Execute(async (e, cancellationToken) =>
       {
                results = await httpClient.GetAsync(..., cancellationToken);
                return true;
       });
```

Synchronous tasks can simply be written like this:
```C#
   schedule
       .AtMinutes(0,15,30,45)
       .Execute(async (e, cancellationToken) =>
       {
                i++;
                return true;
       });
```
or if you don't want to use the async machinery:
```C#
   schedule
       .AtMinutes(0,15,30,45)
       .Execute((e, cancellationToken) =>
       {
                i++;
                return Task.FromResult(true);
       });
```

To handle cancellations you put a try catch around async APIs that take a cancellation token, like this:
```C#
   schedule
       .AtMinutes(0,15,30,45)
       .Execute(async (e, cancellationToken) =>
       {
          try 
          {
                results = await httpClient.GetAsync(..., cancellationToken);
                return true;
          }  
         catch(Exception)
       {
            return false;
       } 
   });
```

Changes are
* Fixed bug in calculation of bit field for years
* Updated unit tests to use years which are compatible (library only allows year dates to be this year - 1)
* Changed signature of IScheduledTask/Execute to return Task<bool> instead of bool.
* Updated unit tests to use async pattern

pettijohn commented:
Neat, I will look at this. I have wanted to add async support to Execute for a long time, just never found the time. I need to look closely at the parenthesis change you proposed in EvaluateRuleMatch(). I admit the software I've written with this always uses * for year, so at least I need to step through unit tests to prove it.

tomlm commented:
The issue was that the (Date - MinYear) needed parens around it, it was applying the & operator to MinYear, not the subtraction between the (Date - MinYear). 
and I needed to fix it for the unit tests to pass.

pettijohn commented:
Yes and hard-coded 2022 year in tests with a min year of 2023 in the core logic would be problematic for running unit tests :) Looks great, I'll publish this soon.

tomlm commented:
Small additional update:
* Added overloads for Execute() and ExecuteAndRetry() that take the old signature.  This makes it backward compatible with the old library.
* Added unit tests for Execute(bool)/Execute(Task<bool>) ExecuteAndRetry(bool)/ExecuteAndRetryAsync(Task<bool>)

Now you can do this
```C#
   // sync scheduled function
   schedule
       .AtMinutes(0,15,30,45)
       .Execute((e, cancellationToken) =>
       {
                i++;
                return true;
       });

   // async scheduled function
   schedule
       .AtMinutes(0,15,30,45)
       .Execute(async (e, cancellationToken) =>
       {
                await Task.Delay(1000);
                return true;
       });
```
The only difference is async() returns a Task<bool> and () returns bool.

pettijohn commented:
Pushed as 24.12.12 to nuget, should be live shortly

pettijohn commented:
I just want to add that I am tickled someone other than me is using this code, and I appreciate you making the library better!
